executor/metric_reader.go

```diff
@@ -0,0 +1,198 @@
+		}
+		rows[i] = row
+	}
+	return rows, nil
+}
+
+func (e *MetricReaderExec) queryMetric(ctx context.Context, queryRange promv1.Range) (pmodel.Value, error) {
+	addr, err := e.getMetricAddr()
+	if err != nil {
+		return nil, err
+	}
+
+	queryClient, err := newQueryClient(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	promQLAPI := promv1.NewAPI(queryClient)
+	ctx, cancel := context.WithTimeout(ctx, promReadTimeout)
+	defer cancel()
+
+	// TODO: add label condition.
+	promQL := e.tblDef.GenPromQL(e.ctx, nil)
+	result, _, err := promQLAPI.QueryRange(ctx, promQL, queryRange)
+	return result, err
+}
+
+func (e *MetricReaderExec) getMetricAddr() (string, error) {
+	// Get PD servers info.
+	store := e.ctx.GetStore()
+	etcd, ok := store.(tikv.EtcdBackend)
+	if !ok {
+		return "", errors.Errorf("%T not an etcd backend", store)
+	}
+	for _, addr := range etcd.EtcdAddrs() {
+		addr = strings.TrimSpace(addr)
+		return addr, nil
+	}
+	return "", errors.Errorf("pd address was not found")
+}
+
+type promQLQueryRange = promv1.Range
+
+func (e *MetricReaderExec) getDefaultQueryRange() promQLQueryRange {
+	return promQLQueryRange{Start: time.Now(), End: time.Now(), Step: time.Second * time.Duration(e.ctx.GetSessionVars().MetricSchemaStep)}
+}
+
+func (e *MetricReaderExec) genRows(value pmodel.Value, r promQLQueryRange) [][]types.Datum {
+	var rows [][]types.Datum
+	switch value.Type() {
+	case pmodel.ValMatrix:
+		matrix := value.(pmodel.Matrix)
+		for _, m := range matrix {
+			for _, v := range m.Values {
+				record := e.genRecord(m.Metric, v, r)
+				rows = append(rows, record)
+			}
+		}
+	}
+	return rows
+}
+
+func (e *MetricReaderExec) genRecord(metric pmodel.Metric, pair pmodel.SamplePair, r promQLQueryRange) []types.Datum {
+	record := make([]types.Datum, 0, 2+len(e.tblDef.Labels)+1)
+	// Record order should keep same with genColumnInfos.
+	record = append(record, types.NewTimeDatum(types.Time{
+		Time: types.FromGoTime(time.Unix(int64(pair.Timestamp/1000), int64(pair.Timestamp%1000)*1e6)),
+		Type: mysql.TypeDatetime,
+		Fsp:  types.MaxFsp,
+	}))
+	record = append(record, types.NewFloat64Datum(float64(pair.Value)))
+	for _, label := range e.tblDef.Labels {
+		v := ""
+		if metric != nil {
+			v = string(metric[pmodel.LabelName(label)])
+		}
+		record = append(record, types.NewStringDatum(v))
+	}
+	if e.tblDef.Quantile > 0 {
+		record = append(record, types.NewFloat64Datum(e.tblDef.Quantile))
+	}
+	return record
+}
+
+type queryClient struct {
+	api.Client
+}
+
+func newQueryClient(addr string) (api.Client, error) {
+	promClient, err := api.NewClient(api.Config{
+		Address: fmt.Sprintf("http://%s", addr),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &queryClient{
+		promClient,
+	}, nil
+}
+
+// URL implement the api.Client interface.
+// This is use to convert prometheus api path to PD API path.
+func (c *queryClient) URL(ep string, args map[string]string) *url.URL {
+	ep = strings.Replace(ep, "api/v1", "pd/api/v1/metric", 1)
```

Comment:
Do we have a chance to set the request header `req.Header.Add("PD-Allow-follower-handle", "true")`?

Reply:
I'm afraid not ...

Reply:
done.